res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js

```diff
@@ -0,0 +1,1336 @@
+  }
+}
+
+TraktorS2MK2.lightDeck = function(group) {
+  // Freeze the lights while we do this update so we don't spam HID.
+  this.freeze_lights = true;
+  for (var packet_name in this.controller.OutputPackets) {
+    packet = this.controller.OutputPackets[packet_name];
+    TraktorS2MK2.lightGroup(packet, group, group);
+    // Shift is a weird key because there's no CO that it is actually associated with.
+    TraktorS2MK2.outputCallback(0, group, "!shift");
+  }
+
+  this.freeze_lights = false;
+  // And now send them all.
+  for (packet_name in this.controller.OutputPackets) {
+    var packet_ob = this.controller.OutputPackets[packet_name];
+    packet_ob.send();
+  }
+}
+
+TraktorS2MK2.init = function(id) {
+  TraktorS2MK2.registerInputPackets();
+  TraktorS2MK2.registerOutputPackets();
+//   var data = [0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x00, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f];
+//   controller.send(data, data.length, 0x80);
+
+  TraktorS2MK2.controller.setOutput("[Master]", "!usblight", 0x7F, true);
+  TraktorS2MK2.lightDeck("[Channel1]");
+  TraktorS2MK2.lightDeck("[Channel2]");
+  TraktorS2MK2.lightDeck("[EffectRack1_EffectUnit1]");
+  TraktorS2MK2.lightDeck("[EffectRack1_EffectUnit2]");
+}
+
+TraktorS2MK2.shutdown = function() {
+  var data = [];
+  for (var i = 0; i < 38; i++) {
+    data[i] = 0;
+  }
+  // Leave USB plug indicator light on.
+  data[0x1C] = 0x7f;
+  controller.send(data, data.length, 0x80);
+  
+  for (i = 0; i < 33; i++) {
+    data[i] = 0;
+  }
+  controller.send(data, data.length, 0x81);
+}
+
+TraktorS2MK2.incomingData = function(data, length) {
+  TraktorS2MK2.controller.parsePacket(data, length);
+}
+
+// The short message handles buttons and jog wheels.
+TraktorS2MK2.shortMessageCallback = function(packet, data) {
+  for (name in data) {
+    field = data[name];
+    if (field.name === "!jog_wheel") {
+      TraktorS2MK2.controller.processControl(field);
+      continue;
+    }
+
+    TraktorS2MK2.controller.processButton(field);
+  }
+}
+
+// There are no buttons handled by the long message, so this is a little simpler.  (Even though
+// this is very similar to the other handler, it's easier to keep them separate to know what's
+// a control and what's a button.
+TraktorS2MK2.longMessageCallback = function(packet, data) {
+  for (name in data) {
+    field = data[name];
+    TraktorS2MK2.controller.processControl(field);
+  }
+}
+
+TraktorS2MK2.samplerGainKnob = function(field) {
+  for (var i = 1; i <= 8; i++) {
+    engine.setParameter("[Sampler" + i + "]", "pregain", field.value / 4096);
+  }
+}
+
+TraktorS2MK2.toggleButton = function(field) {
+  if (field.value > 0) {
+    script.toggleControl(field.group, field.name);
+  }
+}
+
+TraktorS2MK2.shift = function(field) {
+  var group = field.id.split(".")[0];
+  TraktorS2MK2.shift_pressed[group] = field.value;
```

Comment:
the common library has support for shift buttons!

Reply:
I would rather not rely on the library's mysterious magic.

Reply:
the best way to improve a library is to use it and then update/refactor/improve it as needed.  Regardless, I was wrong, there isn't shift support anyway